fpdart v2.0.0

SandroMaglione opened the pull request:
### This PR is an open discussion on what's the best strategy to release fpdart v2 to minimize friction for users of the library

### 👇 Please comment below with any idea, comment, critique or opinion 👇

*** 

## Problems with v1
Too many classes (`IO`, `IOOption`, `IOEither`, `Task`, `TaskOption`, `TaskEither`, `Reader`, `State`, `ReaderTaskEither`):
- Similar implementation with different generic parameters = **A lot of duplicated code** (both core and tests)
```dart
/// [IO] 👇
abstract final class _IOHKT {}

final class IO<A> extends HKT<_IOHKT, A>
    with Functor<_IOHKT, A>, Applicative<_IOHKT, A>, Monad<_IOHKT, A> {
  final A Function() _run;
}

/// [Task] 👇
abstract final class _TaskHKT {}

final class Task<A> extends HKT<_TaskHKT, A>
    with Functor<_TaskHKT, A>, Applicative<_TaskHKT, A>, Monad<_TaskHKT, A> {
  final Future<A> Function() _run; /// 👈 Difference: [Future] here
}
```
- Code duplication = Hard to maintain, more lines of code, more code to read (and understand) for contributors
- Requires conversion between classes (`from*`, `to*`, e.g. `toTask`, `toTaskEither`)
- Requires having a different `Do` constructor for each class, making the do notation harder to use
- Hard to understand for newcomers, hard to reason with and explain for veterans (and verbose)
- More complex code, less contributors

**Too much jargon**: methods and classes are using terms from pure functional programming (math), less clear and hard to understand (e.g. `pure`, `Reader`, `chainFirst`, `traverse`, `Endo`).

Typeclasses do not work well with dart and they cause a lot of overhead to maintain and understand. In fact, they are not necessary to implement the core of fpdart (**they can be removed** 💁🏼‍♂️).

Too many "utility functions" that may be considered outside of the scope of fpdart (e.g. `predicate_extension.dart`).

## fpdart v2 solution: `Effect`
A single `Effect` class that contains the API of all other classes in v1 (similar to `ReaderTaskEither`).

All Effect-classes derive from the same interface `IEffect`:

```dart
abstract interface class IEffect<E, L, R> {
  const IEffect();
  Effect<E, L, R> get asEffect;
}
```

### Benefits
- A lot less code: easier to maintain, contribute, test, understand (a single `effect.dart`)
- No need of conversion methods (a lot less code)
- A single Do notation (implemented as a factory constructor `factory Effect.gen`): the do notation also includes `Option` and `Either` (since both extend `IEffect`)
- No more jargon: easy to understand method names instead of fp jargon (e.g. `succeed` instead of `pure`)
- Removed all typeclasses and unnecessary utility methods
- Easier to explain and understand (focus on learning a single `Effect` and how it works)
- **Smaller API that allows all the same functionalities as before**
- More resources to focus on better documentation, tests, and examples

> **Important**: `Effect` comes from the [`effect`](https://www.effect.website/) library (typescript), which itself was inspired from [`ZIO`](https://zio.dev/).
>
> The `Effect` class and methods in `fpdart` are based on `effect` from typescript (similar API and methods names).
>
> Huge thanks also to @tim-smart for his initial [`zio.dart`](https://github.com/tim-smart/elemental/blob/main/packages/elemental/lib/src/zio.dart) implementation.

### Downsides
-  ⚠️ **Huge breaking change** ⚠️
- Nearly all tests need to be rewritten
- Documentation and examples to redo completely

SandroMaglione committed:
experiment with shared `Effect` class

SandroMaglione committed:
`IEffect` shared interface

SandroMaglione committed:
`Either` as `IEffect`

SandroMaglione committed:
`Async` from `IEffect`

SandroMaglione committed:
`call` run

SandroMaglione committed:
no Sync/Async/Either, just `Effect`

SandroMaglione committed:
`Effect` and `Either` env

SandroMaglione committed:
`Option` as `IEffect`

SandroMaglione committed:
`mapLeft`

SandroMaglione committed:
more `Effect` methods

SandroMaglione committed:
`asEffect` for `IEffect` interface

SandroMaglione committed:
`Effect` api with categories

SandroMaglione committed:
file system example with `Effect`

SandroMaglione committed:
refactor with abstract env

SandroMaglione committed:
test with `provide` in `Effect`

SandroMaglione committed:
`catchError`

SandroMaglione committed:
refactor internals

SandroMaglione committed:
remove all examples
outdated for now with new `Effect`

SandroMaglione committed:
initial (new) tests on `Effect`

SandroMaglione committed:
import alias

SandroMaglione committed:
`Effect` sync execute

SandroMaglione committed:
collect as `Iterable`

SandroMaglione committed:
expand `Effect` API

SandroMaglione committed:
`orDie` API

SandroMaglione committed:
`Option` API

SandroMaglione committed:
`Either` API

SandroMaglione committed:
`Exit` with `Cause`

SandroMaglione committed:
`tryCatch` API

SandroMaglione committed:
removed `Cause` by interrupt

SandroMaglione committed:
`zipLeft` and `zipRight`

SandroMaglione committed:
`forEach` and `all`

SandroMaglione committed:
migration md

SandroMaglione committed:
version 2.0.0-dev.1

SandroMaglione committed:
http request example

SandroMaglione committed:
run `Effect` with `Never` env

SandroMaglione committed:
fetch pokemon with validation example

SandroMaglione committed:
run `gen` as `sync` or `async`

SandroMaglione committed:
pokemon api example with `void` type
https://github.com/dart-lang/language/issues/3670

SandroMaglione committed:
working on types `Never` vs `void`

SandroMaglione committed:
revert to use `Never`
https://github.com/dart-lang/language/issues/3670#issuecomment-2011776343

SandroMaglione committed:
execute either as effect using `Never`

SandroMaglione committed:
Use generic instead of `Never`

SandroMaglione committed:
`onError` for `gen`

SandroMaglione committed:
`catchError` and `catchCause`

SandroMaglione committed:
factory constructors

SandroMaglione committed:
running `Effect`

SandroMaglione committed:
clean up `Cause`

SandroMaglione committed:
fix fail in `Effect.gen`

SandroMaglione committed:
conversions and folding

SandroMaglione committed:
imports

SandroMaglione committed:
prepare v2 dev pre release

SandroMaglione commented:
Released the first pre-release: [fpdart v2.0.0-dev.1](https://pub.dev/packages/fpdart/versions/2.0.0-dev.1)

SandroMaglione committed:
no `Never` env, maybe `void`

airychkov commented:
Hello! Thanks for your mission. 
But I missed  rightsEither(). for me it's good function for safety data load

SandroMaglione committed:
`getSomes`, `getLefts`, `getRights`

SandroMaglione committed:
filtering

airychkov commented:
Thanks!

I have questions. 
Should I rewrite code? Now match only in Effect
I used match with Options and Either
like this

data.match(
() => const Text('not found'),
(final data) => Widget(data: data),
),

SandroMaglione committed:
full `Option` API

tim-smart commented on packages/fpdart/lib/src/effect.dart:
```diff
@@ -0,0 +1,561 @@
+      );
+
+  /// {@category collecting}
+  static Effect<E, L, Iterable<R>> forEach<E, L, R, A>(
```
Concurrency?

SandroMaglione replied:
Working on it, I am thinking how to make it work well with `Iterable` (not just `List`)

tim-smart commented on packages/fpdart/lib/src/effect.dart:
```diff
@@ -0,0 +1,561 @@
+  }
+
+  /// {@category execution}
+  Future<R> runFuture(E env) async {
```
Not sure if providing the env at the end is the best approach, as you might want to eliminate dependencies before running an Effect. One example is `Scope`, if you plan to add that eventually.

Also one big missing piece here is async dependency injection. I attempted to solve this with a simple Layer/Runtime implementation in elemental, but maybe you can come up with something else.

SandroMaglione replied:
I added `provide` and `provideEffect` to provide the full `E` before running:
https://github.com/SandroMaglione/fpdart/blob/a3352b5de8e1c9512ce7abb66834c500fc072f10/packages/fpdart/lib/src/effect.dart#L293-L304

`provideEffect` works for "async dependency", or are you meaning something else?

***

The problem in dart is that you cannot erase part of the dependency, so you must provide all `E`.

Do you have any idea for some way to solve this issue?

tim-smart replied:
No you can't really get around that in dart. Which is why Effect's solution for dependencies isn't really a nice option here.

SandroMaglione committed:
full `Either` API

SandroMaglione committed:
rename `Fail` to `Failure`

SandroMaglione committed:
errors in `FutureOr` sync

SandroMaglione committed:
`onError` with stack trace

SandroMaglione commented on packages/fpdart/lib/src/effect.dart:
```diff
@@ -0,0 +1,561 @@
+    } on Cause<L> {
+      rethrow;
+    } catch (error, stackTrace) {
+      throw Die(error, stackTrace);
```
I implemented `catchAllCause` (called `catchCause`) by making the error channel `Never`:
https://github.com/SandroMaglione/fpdart/blob/f02afe3d4eb4bde8ca45ee769def2b801c0abfa1/packages/fpdart/lib/src/effect.dart#L547-L558

When running then the idea is to `rethrow` `Cause` or throw `Die` otherwise.

What do you think about this approach?

tim-smart replied:
The problem is that you need to catch defects and convert them to Die for every Effect, not just when you call a run* method.

SandroMaglione replied:
I don’t understand this point. Do you mean that there should be a "catch" every time `_unsafeRun` is executed? Could you give me an example where the current approach doesn’t work?

tim-smart replied:
This should return "rescued!"

```dart
import "package:fpdart/fpdart.dart";

void main() {
  final result = Effect<Null, Never, String>.functionSucceed(() {
    throw "fail";
    return "hello!";
  }).catchCause((cause) => Effect.succeed("rescued!")).runSync();

  print(result);
}
```

SandroMaglione replied:
Got it, I [updated the code](https://github.com/SandroMaglione/fpdart/commit/f343b0314c53a45b8c050294b54fff82f64b9761) to create effects from a safe `Effect.from`

SandroMaglione commented on packages/fpdart/lib/src/exit.dart:
```diff
@@ -0,0 +1,53 @@
+
+typedef Exit<L, R> = Either<Cause<L>, R>;
+
+sealed class Cause<L> implements Error {
```
Can you give me an idea of of interruption would work in dart? Is interruption something that the user triggers only manually (`.interrupt`) or is it possible for `Effect` to interrupt by itself in some cases?

tim-smart replied:
You would need to add an option to the async run methods that allow passing in some kind of abort signal (could be a dart Completer). Create a lightweight "handle" to the running Effect would also help, which you could call .interrupt() on.

SandroMaglione replied:
This would be about adding a `Completer` (or `Deferred`) optional parameter to `run*()`.

Then when executing `Effect` the function will first check the value inside `Deferred` and, if already present, return an `Interrupted` and stop execution.

Is my interpretation correct? What are some other usecases that I am missing?

tim-smart replied:
Also it would send a signal to the currently running Effect, so you could hook it up to something like dio's CancelToken.

SandroMaglione replied:
Added interruption using `Deferred` and `Context` (I see now how this all plays together 👌)

SandroMaglione committed:
`provide` and `mapEnv`

SandroMaglione committed:
`provideEffect`

SandroMaglione committed:
test type parameters

SandroMaglione committed:
run only when env is `Null`

SandroMaglione committed:
update examples

SandroMaglione committed:
update examples (again)

SandroMaglione committed:
catch error in gen future

ryanhanks-bestow commented:
Excited to see this in flight.

Question: is there a reason not to match the type signature introduced by the typescript library, in terms of type parameter names and their ordering?

It might be easier to bring the community up-to-speed if you can lean a little harder on the existing effect ecosystem. With the type parameter discrepancies, I find myself getting confused jumping back and forth between the two systems (i.e., is R "Right" or "Requirements"? is E "Error" or effect type?)

I can understand how things ended up here, and maybe there's good reason to keep it this way, given how the type parameters are named in the context of Either.

SandroMaglione committed:
effect abstract only

utamori commented:
I would like to try v2 of this package, but the version constraint of the meta package conflicts with the stable version of Flutter. If there are no issues, how about setting the meta version to 1.11.0?

```
Resolving dependencies...
Note: meta is pinned to version 1.11.0 by flutter_test from the flutter SDK.
See https://dart.dev/go/sdk-version-pinning for details.
Because every version of flutter_test from sdk depends on meta 1.11.0 and fpdart >=2.0.0-dev.1 depends on meta ^1.12.0, flutter_test from sdk is incompatible with fpdart >=2.0.0-dev.1.
```

SandroMaglione committed:
`meta` version

SandroMaglione committed:
run effects with catching

SandroMaglione committed:
`Deferred` and `Context`

SandroMaglione committed:
`race` and `raceAll`

SandroMaglione committed:
`delay` and `timeout`

SandroMaglione committed:
`Deferred` as `Effect`

hbock-42 commented:
Not good enough in fp to review (and understand) the changes but I can give a hand by trying new version before official release if needed.

Yet I agree with your observations regarding `too much jargon`, `Hard to understand for newcomers, hard to reason with and explain for veterans`.

I think a lot of us are already convinced about the advantages of using FP in our code, the problem is about converting others, our coworkers first of all.
So anything making it easier to enter this paradigm is welcomed (not that you have not already done more than the majority), and I think most of us will not mind having to refactor some code.

Thank for your work :)

SandroMaglione committed:
`Deferred` effect with test

SandroMaglione committed:
interruption test

SandroMaglione committed:
provide API

SandroMaglione committed:
fpdart 2.0.0-dev.2

SandroMaglione commented:
New release: [fpdart v2.0.0-dev.2](https://pub.dev/packages/fpdart/versions/2.0.0-dev.2):
- Complete `Option` and `Either` API
- Execute `Effect` using `provide` (with `Null` as dependency)
- Fixed implementation of running `Effect` and catching `Cause`
- Added interruption (`Cause.Interrupted`)
  - `Deferred`
  - `Context`
  - New methods (`raceAll`, `race`, `delay`, `sleep`, `timeout`)

hbock-42 commented on packages/fpdart/lib/src/option.dart:
```diff
@@ -1,617 +1,162 @@
 
-  @override
-  B match<B>(B Function() onNone, B Function(Never t) onSome) => onNone();
+  factory None() => _none;
```
```suggestion
  const factory None() => _none;
```

SandroMaglione committed:
const `None`

SandroMaglione committed:
introduce `Scope`

airychkov commented:
Error stack not so good. Only show main function call, not show when happen

airychkov commented:
I don't know it's right or not
two Fail just for test
if (!sourcePath.isEmpty) {
          Effect.fail(SourceTypeFailure(error: 'empty source'));
}

final sourceType = $.sync(sourcePathType().mapEnv((final env) => sourcePath));

if (!sourcePath.isEmpty) {
     Effect.fail(SourceTypeFailure(error: 'empty source2'));
 }

console log:

flutter: empty source on null
flutter: empty source2 on null

SandroMaglione committed:
`Scope` with test

SandroMaglione committed:
test closable `Scope`

SandroMaglione committed:
v2.0.0-dev.3

SandroMaglione commented:
New release: [fpdart v2.0.0-dev.3](https://pub.dev/packages/fpdart/versions/2.0.0-dev.3):
- Added `Scope`
- `const` constructor for `None`

tim-smart commented on packages/fpdart/lib/src/effect.dart:
```diff
@@ -0,0 +1,800 @@
+  /// {@category constructors}
+  factory Effect.from(FutureOr<Exit<L, R>> Function(Context<E> context) run) =>
+      Effect._((context) {
+        try {
```
You can move __unsafeRun into this constructor and remove it?

tim-smart commented on packages/fpdart/lib/src/effect.dart:
```diff
@@ -0,0 +1,800 @@
+  factory Effect.tryCatch({
+    required FutureOr<R> Function() execute,
+    required L Function(Object error, StackTrace stackTrace) onError,
+    FutureOr<dynamic> Function()? onCancel,
```
Unused

tim-smart commented on packages/fpdart/lib/src/effect.dart:
```diff
@@ -0,0 +1,800 @@
+      );
+
+  /// {@category constructors}
+  factory Effect.functionFail(FutureOr<Cause<L>> Function() f) => Effect.from(
```
functionSucceed/functionFail feel like weird names. Maybe Effect.succeedLazy/failLazy?

tim-smart commented on packages/fpdart/lib/src/effect.dart:
```diff
@@ -0,0 +1,800 @@
+        return deferred.wait<E>().__unsafeRun(context).then(
+              (exit) => signal
+                  .failCause<E, L>(const Interrupted())
+                  .__unsafeRun(context.withoutSignal)
```
Looks like there is a bug, when the parent signal is interrupted it isn't propagated to the child signal.

tim-smart commented on packages/fpdart/lib/src/effect.dart:
```diff
@@ -0,0 +1,800 @@
+
+  /// {@category constructors}
+  static Effect<E, L, void> sleep<E, L>(Duration duration) => Effect.from(
+        (_) => Future.delayed(
```
Maybe dart has a lower level timer api that can be interrupted, to prevent resource leaks.

tim-smart commented on packages/fpdart/lib/src/effect.dart:
```diff
@@ -0,0 +1,800 @@
+  Effect<Null, L, R> provide(Context<E> context) {
+    final env = context.env;
+    final effect = env is ScopeMixin && !env.scopeClosable
+        ? alwaysIgnore(env.closeScope())
```
Hmm this could be an issue, as if there was a defect when try to close the scope, it will be swallowed / unhandled.

tim-smart replied:
You need to close the scope regardless of success or failure.

tim-smart commented on packages/fpdart/lib/src/effect.dart:
```diff
@@ -0,0 +1,800 @@
+
+extension EffectNoScopeFinalizer<E, L, R> on Effect<E, L, R> {
+  /// {@category scoping}
+  Effect<Scope<E>, L, R> get withScope => Effect<Scope<E>, L, R>.from(
```
You can remove this, as really you should just use .addFinalizer directly.

tim-smart commented on packages/fpdart/lib/src/effect.dart:
```diff
@@ -0,0 +1,800 @@
+  Effect<Scope<E>, L, R> acquireRelease(
+    Effect<Null, Never, Unit> Function(R r) release,
+  ) =>
+      withScope.tapEnv(
```
Technically you should make the resource acquisition uninterruptible here, would require some thought about how to make interruption toggle-able / regional.

ryanhanks-bestow commented:
Hey Sandro.

Looking good, I've been playing around.

Given
```
typedef ProduceEffect<State, ProductionError, Event>
    = Effect<State, ProductionError, Iterable<Event>>;
typedef HandleEffect<State, HandlerError, Event>
    = Effect<(State, Event), HandlerError, State>;
typedef ApplyEffect<State,Error,Event> = Effect<State, Error, (Iterable<Event>,State)>
```, how would you do a fold across the iterable result of a ProduceEffect using a HandleEffect to result in an Effect of type ApplyEffect?

Feel free to suggest changes in the shapes of those definitions if you have a better starting approach.

Thanks

ryanhanks-bestow commented:
> Hey Sandro.
> 
> Looking good, I've been playing around.
> 
> Given
> 
> ```
> typedef ProduceEffect<State, ProductionError, Event>
>     = Effect<State, ProductionError, Iterable<Event>>;
> typedef HandleEffect<State, HandlerError, Event>
>     = Effect<(State, Event), HandlerError, State>;
> typedef ApplyEffect<State,Error,Event> = Effect<State, Error, (Iterable<Event>, State)>
> ```
> 
> , how would you do a fold across the iterable result of a ProduceEffect using a HandleEffect to result in an Effect of type ApplyEffect?
> 
> Feel free to suggest changes in the shapes of those definitions if you have a better starting approach.
> 
> Thanks

Also, curious if you have any general suggestions in naming type alias of an effect. Those names aren't the names I'm using verbatim, but I have been applying a naming convention where a prefix an alias with a verb matching the general action that maps an Env value to the result and error it could produce. So for example, in the actual code I'm working on, I've implemented typedefs as such:
```dart
typedef DispatchEffect<State, DispatchError, Event>
    = Effect<State, DispatchError, Iterable<Event>>;

typedef StateEventHandlerEffect<State, HandlerError, Event>
    = Effect<(State,Event), HandlerError, State>;
```
Again, not 100% sure I'm shaping things here, maybe with this you'll be able to let me know if I'm applying the general use correctly.

Thanks

ryanhanks-bestow commented:
@SandroMaglione what editor do you use? In intelliJ, when I'm creating an Effect using `Effect.gen`, I lose type-awareness inside the call to sync (even if I explicitly supply the type parameter A).

Wondering if you might consider this alternative definition for EffectGen:
```
class EffectGen<E, L> {
  EffectGen({
    required FutureOr<A> Function<A>(IEffect<E, L, A>) async,
    required A Function<A>(IEffect<E, L, A>) sync,
  })
      : this.sync = sync,
        this.async = async;

  FutureOr<A> async<A>(IEffect<E, L, A>effect) => this.async(effect);

  A sync<A>(IEffect<E, L, A>effect) => this.sync(effect);

}
```

SandroMaglione committed:
refactor and remove `__unsafeRun`

SandroMaglione committed:
removed `tryCatch` `onCancel`

SandroMaglione committed:
renamed `succeedLazy` and `failLazy`

SandroMaglione commented on packages/fpdart/lib/src/effect.dart:
```diff
@@ -0,0 +1,800 @@
+      });
+
+  /// {@category constructors}
+  static Effect<E, L, void> sleep<E, L>(Duration duration) => Effect.from(
+        (_) => Future.delayed(
+          duration,
+          () => const Right(null),
+        ),
+      );
```
It is possible to use `Timer`+`Completer`.

@tim-smart How would interruption work for `sleep`? How can `sleep` detect an interrupt and block timer?

```suggestion
  static Effect<E, L, void> sleep<E, L>(Duration duration) => Effect.from(
        (_) {
          final completer = Completer<Exit<L, void>>();
          Timer(duration, () {
            completer.complete(const Right(null));
          });

          return completer.future;
        },
      );
```

tim-smart replied:
First you will want to add a Effect.async/asyncInterrupt for callback apis, then use Effect.asyncInterrupt with the Timer api. Then hook up timer.cancel.

SandroMaglione replied:
@tim-smart [I added `async`](https://github.com/SandroMaglione/fpdart/pull/147/commits/1118bd66039da1ac926b178ab347be92117b5df8), but I am having some issues understanding the mental model for `Deferred` and interruption.

Could you give me an example of how to manually interrupt an `Effect` with code? Does this required `Deferred` or is it about [`Context`](https://github.com/SandroMaglione/fpdart/blob/effect-refactor/packages/fpdart/lib/src/context.dart)?

tim-smart replied:
Sleep should be:

```dart
  static Effect<E, L, void> sleep<E, L>(Duration duration) =>
      Effect.asyncInterrupt(
        (resume) {
          final timer = Timer(duration, () {
            resume.succeed(null);
          });
          return Effect.succeedLazy(() {
            timer.cancel();
          });
        },
      );
```

tim-smart replied:
As for interrupting an Effect, here is some pseudo code:

```dart
final cancel = Effect.never.runCallback();

// some time later...

cancel();
```

SandroMaglione replied:
@tim-smart what is `runCallback` in the example? What is the equivalent in `Effect`?

Is `cancel` a `Deferred`? Should there be any API that returns a function to interrupt the `Effect`?

tim-smart replied:
https://effect-ts.github.io/effect/effect/Effect.ts.html#runcallback

In dart maybe something like:

```dart
void Function() runCallback({ void Function(Exit<E, A> exit)? onExit })
```

SandroMaglione replied:
Is `runCallback` the only way to manually trigger an interruption?

How would you test interruption with `sleep` (or also `asyncInterrupt`)?

tim-smart replied:
In Effect (typescript) you can pass an abort signal to Effect.runPromise, use the cancel callback from Effect.runCallback or call interrupt on fiber returned from Effect.runFork.

dart doesn't have a built in "cancellation" primitive, so you could use a Future potentially.

SandroMaglione replied:
Do you think is it possible to use `Deferred` (or `Context`) in fpdart to give access to interruption?

tim-smart replied:
Not too sure what you mean, might just have to try it out!

SandroMaglione committed:
scope methods in `Effect`

SandroMaglione committed:
fix `withScope` issue

SandroMaglione commented:
@ryanhanks-bestow could you provide a screenshot of IntelliJ that shows the type not inferred? Do you know of any limitation of IntelliJ regarding Records?

SandroMaglione committed:
`async`, `asyncInterrupt`, `sleep`

SandroMaglione committed:
rename constructors sync/future

SandroMaglione committed:
example read/write file

SandroMaglione committed:
`sleep`

lishaduck commented:
Hello! Is there anything I could do to help make some progress on this? I'd really like to switch some code to FPDart, but I don't want to if it'll soon need to be rewritten. In addition, I really like Effect-TS and would love to see a Dart equivalent.
I'm certainly happy to help with tests and docs, and would be willing to work through porting implementations.

SandroMaglione commented:
> Hello! Is there anything I could do to help make some progress on this? I'd really like to switch some code to FPDart, but I don't want to if it'll soon need to be rewritten. In addition, I really like Effect-TS and would love to see a Dart equivalent. I'm certainly happy to help with tests and docs, and would be willing to work through porting implementations.

The initial implementation is there. The main issue now is with composition. The dart language doesn’t support union types and type inference is limited. This means that you are required to specific manually all types, which is not ideal from a developer UX perspective. So much so that it doesn’t yet justify using fpdart's `Effect` over usual OOP dart patterns.

This may be solved by static metaprogramming, so I am actually waiting for some progress on the language.

lishaduck commented:
> The initial implementation is there. The main issue now is with composition. … This may be solved by static metaprogramming, so I am actually waiting for some progress on the language.

Ok then. Thanks for the explanation. Would expect fpdart 2 to be stable enough to use then, just with the expectation that it'll get more concise in the future, or do you anticipate macros to entirely subsume the current API?

SandroMaglione commented:
> > The initial implementation is there. The main issue now is with composition. … This may be solved by static metaprogramming, so I am actually waiting for some progress on the language.
> 
> Ok then. Thank you for the explanation, it's very helpful for understanding the current progress. Would you expect fpdart 2 to be stable enough to use then, just with the expectation that it'll get more concise in the future, or do you anticipate macros to entirely subsume the current API?

I don't expect the `Effect` encoding to change. Nonetheless, static metaprogramming has the potential to disrupt many common dart patterns. I would advise against using fpdart v2 in production yet.

That being said, the API as it is now works and hopefully won't need to change too much.

lishaduck commented:
> I don't expect the `Effect` encoding to change ... That being said, the API as it is now works and hopefully won't need to change too much.

Ok, thanks!

> Nonetheless, static metaprogramming has the potential to disrupt many common dart patterns. I would advise against using fpdart v2 in production yet.

It's primarily for a personal project, so there's not really a "production". I guess I'll give it a go and report back with any feedback. Thanks again for fpdart!

lishaduck commented:
Finally got some time to work on migrating some code to FPDart.
Feedback:
- I miss [`tryPromise`](https://effect.website/docs/guides/essentials/creating-effects#trypromise)
- I haven't used [`Cause`](https://effect.website/docs/other/data-types/cause) much, but if there's not a `tryPromise`, I might more.
- A lack of row polymorphism for records is annoying? I thought I'd found a good way to type context, but [it intentionally doesn't work](https://github.com/dart-lang/language/issues/1291#issuecomment-726421985).

P.S.: I'll update this comment if I have anything else to add instead of notifying everyone.

lishaduck commented on packages/fpdart/lib/src/option.dart:
```diff
@@ -1,617 +1,159 @@
-final class _OptionThrow {
-  const _OptionThrow();
-}
+part of "effect.dart";
```
~~Would a micro module be possible, like [Effect's](https://effect.website/docs/other/micro), but even lighter?
In my head, you could keep `IEffect` as base, but keep `Option`, `Either`, etc from directly interacting with `Effect`, but only `IEffect`. That would make migrating from v1 and integrating it into existing projects easier, by not requiring users to use the Effect system (it's obviously recommended to use the Effect portion), and comes with the side-effect of a (slightly) smaller learning curve, as you can use it in chunks.~~

I thought there were some APIs here that called Effect. Doesn't look like anything really, so I think it should be documented that you can still use `Option`, `Either`, etc, without needing to adopt the Effect system.

lishaduck commented on packages/fpdart/lib/src/exit.dart:
```diff
@@ -0,0 +1,71 @@
+
+typedef Exit<L, R> = Either<Cause<L>, R>;
+
+sealed class Cause<L> implements Error {
```
Oh, also, could this get its own module? I get that it's an internal detail, but `Exit` feels different than `Cause`. Maybe rename it rather than splitting it given the size of `Exit`?

lishaduck commented on packages/fpdart/lib/src/effect.dart:
```diff
@@ -0,0 +1,843 @@
+      );
+
+  /// {@category folding}
+  Effect<E, Never, C> match<C>({
```
This has a semi-non-trivial definition, so I'm hesitant to suggest this, but maybe it'd be possible to use native pattern matching? i.e.

```dart
final val = switch ($.syncEither(effect)) {
  Left() => ...,
  Right() => ...,
}
```

rather than

```dart
final val = effect.match(
  () => ...,
  () => ...,
)
```

Yes, it's more verbose, but the native pattern matching feels much more "right" to me. The function feels like it's working around an issue from JS that isn't in Dart.

lishaduck reviewed (commented):
I left some comments, feel free to disregard.

lishaduck commented on packages/fpdart/lib/src/either.dart:
```diff
@@ -1,675 +1,214 @@
-        Foldable2<_EitherHKT, L, R>,
-        Alt2<_EitherHKT, L, R>,
-        Extend2<_EitherHKT, L, R> {
+sealed class Either<L, R> extends IEffect<Never, L, R> {
```
I've thought about this more, and I think keeping the generics the same as IEffect's makes sense.

airychkov commented:
Hello! My Second try to use Effect. 

test code

abstract interface class FileLoader<T> {
  Effect<ConvertFromJson<T>, DataError, IList<T>> loadSource(final String source);
}

final class JsonFileLoader<T> implements FileLoader<T> {
  @override
  loadSource(final String source) => Effect.gen(
        (final $) async {
          final fileData = await $.async(Effect.tryCatch(
              execute: () => rootBundle.loadString(source, cache: true),
              onError: (final error, final stackTrace) => FileError(error: error, stackTrace: stackTrace)));

          return const IList.empty();
        },
      );
}

call Effect 
final c = await b.loadSource('asset/file.yaml').provideEnv(File.fromJson).runFutureExit();

get exception 

Exception has occurred.
FlutterError (Unable to load asset: "asset/file.yaml".
The asset does not exist or has empty data.

I expect Either.Left If I right undertund examples. 
Most of examples is outdated now

lishaduck commented:
> Hello! My Second try to use Effect.
> 
> test code
> 
> ```dart
> abstract interface class FileLoader { Effect<ConvertFromJson, DataError, IList> loadSource(final String source); }
> 
> final class JsonFileLoader implements FileLoader { @OverRide loadSource(final String source) => Effect.gen( (final $) async { final fileData = await $.async(Effect.tryCatch( execute: () => rootBundle.loadString(source, cache: true), onError: (final error, final stackTrace) => FileError(error: error, stackTrace: stackTrace)));
> 
>       return const IList.empty();
>     },
>   );
> 
> }
> 
> call Effect final c = await b.loadSource('asset/file.yaml').provideEnv(File.fromJson).runFutureExit();
> ```
> 
> get exception
> 
> Exception has occurred. FlutterError (Unable to load asset: "asset/file.yaml". The asset does not exist or has empty data.
> 
> I expect Either.Left If I right undertund examples. Most of examples is outdated now

I'm unsure as well, the code should catch it, but if you look at [the v2 examples](https://github.com/SandroMaglione/fpdart/tree/effect-refactor/examples), you'll see that they call `catchError`. I think that'd work work around it.
Given that it's a `FlutterError`, perhaps it's got something to do with the frameworks' [weird error handling behaviors](https://docs.flutter.dev/testing/errors)? Otherwise, perhaps the `Effect` doesn't handle `Error`s, just `Exceptions`, though it looks like it does to me.

Disclaimer: #155 discouraged me enough to wait for macros, so I haven't written much with fpdart in a few months.

harkairt commented:
How does the fact that [Static Metaprogramming](https://github.com/dart-lang/language/issues/1482) is currently iceboxed / `state-reject` affects the future of fpdart 2.0?

SandroMaglione commented:
> How does the fact that [Static Metaprogramming](https://github.com/dart-lang/language/issues/1482) is currently iceboxed / `state-reject` affects the future of fpdart 2.0?

Hi everyone,

An update on the state of fpdart and fpdart v2. [Static Metaprogramming is `state-rejected`](https://github.com/dart-lang/language/issues/1482), therefore it will not be added to Dart. This introduces a problem with fpdart v2.

The new `Effect` API requires to manually specify all the dependencies for each execution (e.g. `Effect<Env, PokemonError, Pokemon>` from the [PR example](https://github.com/SandroMaglione/fpdart/blob/29331c829b0f45af50b32342a39d89cb9b4a192c/examples/poke_api/lib/main.dart#L29)). That’s because it’s not possible in Dart to infer all the types (or at least I didn’t find a solution for this).

Without inferred types, the `Effect` API is arguably not much usable and becomes quickly verbose. 

There doesn’t seem to be [any upcoming feature](https://github.com/orgs/dart-lang/projects/90/views/1) that may solve this issue.

My plan would be to keep fpdart v2 as pre-release, waiting for some language improvement that solves the infer problem (or someone coming up with an alternative solution).

In the meanwhile, I am **not** planning in adding any more features or API to the stable version of fpdart. I am open for PRs if anyone is interested in contributing with improvements and bug fixes.

Open to discuss this if anyone has any better idea or suggestion 👇

SandroMaglione commented:
@lucavenir the two main issues are:
1. Dart cannot infer nested requirements for `Effect` ([the third type parameter `R`](https://github.com/SandroMaglione/fpdart/blob/29331c829b0f45af50b32342a39d89cb9b4a192c/packages/fpdart/lib/src/effect.dart#L65))
2. Dart doesn’t support "dynamic" union types (e.g. `A | B`), which are necessary to express a list of multiple requirements

Code generation may help. You could inspect the code and generate a [`sealed` class](https://dart.dev/language/class-modifiers#sealed) (acting as a union type) that collects all the requirements that the inspection found by recursively checking your code. That's an option that I foresee possible, although it may not be neither easy to implement nor a great developer experience.

RandalSchwartz commented:
The updated analyzer plugin interface might help.  People are reportedly using it for realtime "build-runner" tooling.
